src/cmdlib.sh

```diff
@@ -127,10 +144,105 @@ EOF
     fi
 
     rm -f ${changed_stamp}
-    set -x
-    sudo rpm-ostree compose tree --repo=${workdir}/repo-build --cachedir=${workdir}/cache \
-         --touch-if-changed "${changed_stamp}" \
-         ${treecompose_args} \
-         ${TREECOMPOSE_FLAGS:-} ${manifest} "$@"
-    set +x
+
+    # really, we don't need to do this whole jazz for `fetch`, but meh, it's
+    # just some minimal I/O in that case
+    cat > ${TMPDIR}/compose.sh <<EOF
+SUDO=
+if [ \$(id -u) != 0 ]; then
+    SUDO=sudo
+fi
+
+# build in a temp bare-user repo in cache/ so we can hardlink from pkgcache
+\$SUDO ostree --repo=${workdir}/cache/repo-build init --mode=bare-user
+
+# seed with commit metadata of last build
+# TODO: don't rely on ${ref}
+if ostree --repo=${workdir}/repo rev-parse ${ref} 2>/dev/null; then
+    \$SUDO ostree --repo=${workdir}/cache/repo-build pull-local \
+        ${workdir}/repo ${ref} --commit-metadata-only
+fi
+
+\$SUDO rpm-ostree compose tree --repo=${workdir}/cache/repo-build \
+    --cachedir=${workdir}/cache --touch-if-changed "${changed_stamp}" \
+    ${treecompose_args} ${TREECOMPOSE_FLAGS:-} \
+    ${manifest} $@
+
+\$SUDO chmod -R a+rX ${workdir}/cache/repo-build/objects
+# can't use ${changed_stamp}; it has different semantics with --download-only
+if ostree --repo=${workdir}/cache/repo-build rev-parse ${ref} 2>/dev/null; then
+    ostree --repo=${workdir}/repo pull-local ${workdir}/cache/repo-build ${ref}
+fi
+
+# just nuke it instead of pruning, it's fast to rebuild from pkgcache hardlinks
+\$SUDO rm -rf ${workdir}/repo-build
+EOF
+
+    # this is the heart of the privs vs no privs dual path
+    if has_privileges; then
+        . ${TMPDIR}/compose.sh
+    else
+        runvm bash ${TMPDIR}/compose.sh
+    fi
+}
+
+runvm() {
+    local vmpreparedir=${workdir}/tmp/supermin.prepare
+    local vmbuilddir=${workdir}/tmp/supermin.build
+
+    # use REBUILDVM=1 if e.g. hacking on rpm-ostree/ostree and wanting to get
+    # the new bits in the VM
+    if [ ! -f ${vmbuilddir}/.done ] || [ -n "${REBUILDVM:-}" ]; then
+        rm -rf ${vmpreparedir} ${vmbuilddir}
+        mkdir -p ${vmpreparedir} ${vmbuilddir}
+
+        local rpms=
+        # then add all the base deps
+        for dep in $(grep -v '^#' /usr/lib/coreos-assembler/vmdeps.txt); do
+            rpms+="$dep "
+        done
+
+        supermin --prepare --use-installed $rpms -o "${vmpreparedir}"
+        cat > "${vmpreparedir}/init" <<EOF
+#!/bin/bash
+set -xeuo pipefail
+workdir=${workdir}
+$(cat /usr/lib/coreos-assembler/supermin-init-prelude.sh)
+rc=0
+sh ${TMPDIR}/cmd.sh || rc=\$?
+echo $rc > ${workdir}/tmp/rc
+/sbin/reboot -f
+EOF
+        chmod a+x ${vmpreparedir}/init
+        (cd ${vmpreparedir} && tar -czf init.tar.gz --remove-files init)
+        supermin --build "${vmpreparedir}" --size 5G -f ext2 -o "${vmbuilddir}"
+        touch "${vmbuilddir}/.done"
+    fi
+
+    echo "$@" > ${TMPDIR}/cmd.sh
```

Comment:
in this case this works because `${TMPDIR}` is under `${workdir}` right?

I've actually seen issues running in openshift where having $TMPDIR be on the nfs mounted volume caused real issues for me so I changed it to not be exported: https://github.com/dustymabe/coreos-assembler/commit/7ecba34d2be7037bd14144d7621e0c5e0920d2a6